### ISSUE TITLE
[DistributionLoss] Add a "return samples" option

### ISSUE THREAD

**Issue by Antoine-Schwartz** [Nixtla/neuralforecast#864]
### Description

As already proposed by [DeepAR on Sagemaker](https://docs.aws.amazon.com/sagemaker/latest/dg/deepar-in-formats.html#deepar-json-response), it would be interesting to be able to retrieve the output samples when using a `DistributionLoss`.

### Use case

Retrieving these trajectories directly in outputs allows for greater independence of use cases that come after forecasting. Typically, being able to recalculate different quantiles without having to relaunch a forecast job is a must for many supply chain optimization modules.

**Comment by elephaint** [comment]
Hi, apologies for the late reply. You can achieve what you want by specifying a large number of quantiles, for example:

`loss = DistributionLoss(distribution='StudentT', quantiles=np.linspace(0, 1, 100))
`

This will effectively allow you to obtain the sampled values for the distribution per timestep in the forecast.

Does that answer your question?

**Comment by carlocav** [comment]
Hello! Is this still the only way to get sample paths? The problem of this approach is that it only works to get the marginal distribution and doesn't allow to get the joint forecast distribution at multiple future time periods taking temporal autocorrelation into account.

**Comment by elephaint** [comment]
As of 3.0.0, you can specify a quantiles argument in predict, i.e.

nf.predict(...., quantiles=[0.1, 0.2,....])

this works for DistributionLosses and point losses with conformal intervals.

**Comment by carlocav** [comment]
Thank you for the answers. This still is returning quantiles rather than actual sample paths over time to approximate the joint forecast distribution.

Sample paths are generated so to compute quantiles and it would be nice to add the option to simply return them. This feature is available in gluonts library and it’s probably the only reason to keep using it.

**Comment by elephaint** [comment]
Sure, we could add a sample endpoint, don't think that's too hard. I'll reopen the issue.

Although not sure if you sample and do it e.g. 200 times to compute a distribution, you get the same as just outputting the quantiles. I mean I somewhat understand the desire to avoid max one line of additional code but it's also just one line? Or am I misunderstanding what it is that you seek?

**Comment by carlocav** [comment]
distributionloss generates samples paths that are then used by neurlaforecast to compute and return quantiles. What I seek is the option to get them directly rather than the quantiles. In fact if we have a forecast horizon > 1 and we want to approximate the joint distribution over time we need the raw sample paths as we can’t use the quantiles of outcomes at single points in time because of temporal autocorrelation e.g if we have a very high value in t+1 it might be likely to have it also in t+2, etc. If I have sample paths I can compute all the quantities of interest, as an example we can compute quantiles of the cumulative sum of the outcomes in t+1 , … t + n adding up the quantities of time 1 to n of each sample and than compute the quantiles, and it would not be possible to compute this quantity  by adding the quantiles at the n points in time because quantiles are not additive. 

You can find other possible use cases where sample paths rather than quantiles are needed e.g. here https://minimizeregret.com/post/2022/07/25/use-sample-paths/

**Comment by elephaint** [comment]
> distributionloss generates samples paths that are then used by neuralforecast to compute and return quantiles. What I seek is the option to get them directly rather than the quantiles. In fact if we have a forecast horizon > 1 and we want to approximate the joint distribution over time we need the raw sample paths as we can’t use the quantiles of outcomes at single points in time because of temporal autocorrelation e.g if we have a very high value in t+1 it might be likely to have it also in t+2, etc. If I have sample paths I can compute all the quantities of interest, as an example we can compute quantiles of the cumulative sum of the outcomes in t+1 , … t + n adding up the quantities of time 1 to n of each sample and only then compute the quantiles, and it would not be possible to compute this quantity by adding the quantiles at the n points in time directly because quantiles are not additive.
> 
> You can find other possible use cases where sample paths rather than quantiles are needed e.g. here https://minimizeregret.com/post/2022/07/25/use-sample-paths/

But that's not how any of the NF models normally work. The forecasts are direct, i.e. there is no such thing as a 'sample path'. In case of a `DistributionLoss`, every model (except for ones that support recursive predictions) will generate forecasts directly. This means, for a forecast horizon _h_, and (e.g.) a Normal distribution, the output is a matrix of size [batch_size, horizon, 2], where the last dimension contains the mean and standard deviation per horizon. When sampling, we sample for each point in the batch and horizon individually using the mean and std per horizon and batch. 

Consequently, there is no difference in sampling the entire horizon path using `Normal().sample()` or just sampling a random quantile in `.predict`. What you are referring to is doing recursive forecasts, which we support for a few selected models (LSTM, RNN, GRU), but which I'd generally avoid as direct forecasts are nearly always better, as they don't propagate bias and variance.

If we were to implement a `.sample` endpoint you get the exact same results as sampling a random quantile. In fact, I'll most likely implement it simply by calling the predict endpoint with a random quantile.

**Comment by carlocav** [comment]
Yes exactly. That would be awesome, as a starting point I think with DeepAR https://nixtlaverse.nixtla.io/neuralforecast/models.deepar.html you already generate quantiles from sample paths/trajectories isn’t it? So here it should be straightforward to return them as it is done in the gluonts library. Then of course it would be great to have alternatives to DeepAR to generate the joint forecast distribution.

**Comment by nic-hue** [comment]
Hello @elephaint looking at the documentation of DeepAR - Nixtla it looks like you are using sampling trajectories/paths to produce the quantiles rather than marginal distributions, is that correct? If so, do you consider in the future the add to the predict method the option to get the raw predicted trajectories when using deepAR models?
